lib/fencing/st_output.c

```diff
@@ -33,6 +33,89 @@ time_t_string(time_t when) {
     return buf;
 }
 
+/*!
+ * \internal
+ * \brief Return a status-friendly description of fence history entry state
+ *
+ * \param[in] history  Fence history entry to describe
+ *
+ * \return One-word description of history entry state
+ * \note This is similar to stonith_op_state_str() except user-oriented (i.e.
+ *       for cluster status) instead of developer-oriented (for debug logs).
+ */
+static const char *
+state_str(stonith_history_t *history)
+{
+    switch (history->state) {
+        case st_failed: return "failed";
+        case st_done:   return "successful";
+        default:        return "pending";
+    }
+}
+
+/*!
+ * \internal
+ * \brief Create a description of a fencing history entry for status displays
+ *
+ * \param[in] history          Fencing history entry to describe
+ * \param[in] full_history     Whether this is for full or condensed history
+ * \param[in] later_succeeded  Node that a later equivalent attempt succeeded
+ *                             from, or NULL if none
+ *
+ * \return Newly created string with fencing history entry description
+ *
+ * \note The caller is responsible for freeing the return value with g_free().
+ * \note This is similar to stonith__event_description() except for
+ *       history entries in status displays rather than event notifications in
+ *       log messages.
+ */
+gchar *
+stonith__history_description(stonith_history_t *history, bool full_history,
+                             const char *later_succeeded)
+{
+    GString *str = g_string_sized_new(256); // Generous starting size
+    char *retval = NULL;
+
+    g_string_printf(str, "%s of %s %s",
+                    stonith_action_str(history->action), history->target,
+                    state_str(history));
+
+    // For failed actions, add exit reason if available
+    if ((history->state == st_failed) && (history->exit_reason != NULL)) {
+        g_string_append_printf(str, " (%s)", history->exit_reason);
+    }
+
+    g_string_append(str, ": ");
+
+    // For completed actions, add delegate if available
+    if (((history->state == st_failed) || (history->state == st_done))
+        && (history->delegate != NULL)) {
+        g_string_append_printf(str, "delegate=%s, ", history->delegate);
+    }
+
+    // Add information about originator
+    g_string_append_printf(str, "client=%s, origin=%s",
+                           history->client, history->origin);
+
+    // For completed actions, add completion time
+    if ((history->state == st_failed) || (history->state == st_done)) {
+        char *completed_time = time_t_string(history->completed);
+
+        g_string_append(str, (full_history? ", completed" : ", last-failed"));
+        g_string_append_printf(str, "='%s'", completed_time);
```

Comment:
I think the 'last-failed' vs 'last-successful' logic disappeared from here.

Reply:
The technical output is unchanged, like "... failed/successful: ... completed/last-failed=TIME". If full_history is false, we're only showing the most recent failure for each node, and I think "last-failed" is just meant to emphasize that.

For the friendly output, state_str() is immediately before this as e.g. "failed at $TIME" or "successful at $TIME". I suppose if full_history is false, we could make it "last failed at $TIME" instead.